staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go

```diff
@@ -264,3 +300,65 @@ func v4WriteStatusFunc(stream io.Writer) func(status *apierrors.StatusError) err
 		return err
 	}
 }
+
+func newStreamExecutor(config *rest.Config, method string, url *url.URL) (*streamExecutor, error) {
+	wrapper, upgradeRoundTripper, err := spdytransport.RoundTripperFor(config)
+	if err != nil {
+		return nil, err
+	}
+	return &streamExecutor{
+		upgrader:  upgradeRoundTripper,
+		transport: wrapper,
+		method:    method,
+		url:       url,
+		protocols: []string{
+			remotecommandconsts.StreamProtocolV4Name,
+			remotecommandconsts.StreamProtocolV3Name,
+			remotecommandconsts.StreamProtocolV2Name,
+			remotecommandconsts.StreamProtocolV1Name,
+		},
+	}, nil
+}
+
+func TestStreamExitsAfterConnectionIsClosed(t *testing.T) {
+	options := StreamOptions{
+		Stdin:  &fakeEmptyDataPty{},
+		Stdout: &fakeEmptyDataPty{},
+	}
+	server := newTestHTTPServer(fakeMassiveDataAttacher, &options)
+
+	ctx, cancelFn := context.WithTimeout(context.Background(), 500*time.Millisecond)
+	defer cancelFn()
+
+	uri, _ := url.Parse(server.URL)
+	exec, err := newStreamExecutor(&rest.Config{Host: uri.Host}, "POST", uri)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	conn, streamer, err := exec.newConnectionAndStream(ctx, options)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	started := make(chan struct{})
+	errorChan := make(chan error)
+	go func() {
+		started <- struct{}{}
+		errorChan <- streamer.stream(conn)
+	}()
+
+	// Wait until stream goroutine starts.
+	<-started
+	time.Sleep(100 * time.Millisecond)
```

Comment:
Is the purpose of this sleep to give time to the underlying SPDY connection to establish new streams? I guess without this sleep, the streamer can be stuck at this 30s [Wait](https://github.com/kubernetes/kubernetes/blob/d90342e500a43e5112da04ecefa0c9ae1b1fb6c2/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/connection.go#L141) when `conn.Close()` is called. And because this 30s Wait is uninterruptible, then the test will fail because the `streamer.stream()` will not be cancelled within 1s.

Relying on the 100ms sleep can be flaky. Instead, can we read/write some contents through the streams to indirectly verify that the streams have been established?

Reply:
No, the sleep is to avoid connection being closed before the stream, which is a false negative. But yes we can have a method to indicate if the stream starts by reading from the stream.

Reply:
> No, the sleep is to avoid connection being closed before the stream, which is a false negative.

Yeah, that's also possible.